src/vmm/src/device_manager/mmio.rs

```diff
@@ -78,46 +81,92 @@ pub struct MMIODeviceInfo {
     pub irqs: Vec<u32>,
 }
 
+struct IrqManager {
+    #[cfg(target_arch = "x86_64")]
+    first: u32,
+    last: u32,
+    next_avail: u32,
+}
+
+impl IrqManager {
+    pub fn new(first: u32, last: u32) -> Self {
+        Self {
+            #[cfg(target_arch = "x86_64")]
+            first,
+            last,
+            next_avail: first,
+        }
+    }
+
+    pub fn get(&mut self, count: u32) -> Result<Vec<u32>> {
+        if self.next_avail + count > self.last + 1 {
+            return Err(Error::IrqsExhausted);
+        }
+        let mut irqs = Vec::with_capacity(count as usize);
+        for _ in 0..count {
+            irqs.push(self.next_avail);
+            self.next_avail += 1;
+        }
+        Ok(irqs)
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    pub fn check(&self, irqs: &[u32]) -> Result<()> {
```

Comment:
Nit: `check()` is very vague. I would rename this function to something like `are_irqs_valid()` or `are_irqs_available()`